engine/random.cpp

```diff
@@ -0,0 +1,42 @@
+#include "random.h"
+
+namespace pong
+{
+
+static std::random_device rd;
+static std::mt19937 generator(rd());
+
+template<typename T>
+T Random::Range(T min, T max)
+{
+    static_assert(false);
+}
+
+template<>
+int Random::Range<int>(int min, int max)
+{
+    std::uniform_int_distribution<int> dist(min, max);
+    return dist(generator);
+}
+
+template<>
```

Comment:
can you explain to me (because I don't know) why the template braces are empty here? What is the reason for it?

Reply:
Well I am not entirely certain myself, at least not past "it needs to be empty otherwise it wont compile." But I guess when using template specialization, since you are not using the generic T for the type name, there is no need to specify the type in the braces since you doing it later like here with `Range<float>` 
I do think it is helpful having it, since you learn at a glance that this is a template function, but it has just been specialized

Reply:
Seems reasonable to me!!


From Dr. GPT:


GPT

The feature you're referring to is called "template specialization" in C++. In the context of your code snippet, it's used to define a specific implementation of a template function for a particular data type, in this case, int.

Here's a brief explanation:

    Template Specialization: This is a C++ feature that allows you to define a specific implementation of a template class or function for a particular data type or a set of data types. It's useful when the general template implementation is not suitable or needs to be optimized for certain types.

In your example, template<> indicates that this is a specialization of the Random::ValueInRange template function for the type int. Because it's a specialization, you don't need to specify the template parameter (<int>) again in the function definition; it's implied by the context.

To learn more about this feature, you could research "C++ template specialization."

For your next turn, you might want to ask about specific use cases or examples of template specialization.